### ISSUE TITLE
Strange behaviour when changing par.G (lecture 2)

### ISSUE THREAD

**Issue by AntonSteino** [ThomasHJorgensen/HouseholdBehaviorCourse#1]
Hi, I've looked at the lecture 2 notebook, playing around to introduce a handheld "retirement" period by giving G a time subscript. This results in a strange overconsumption pattern in the last period pre retirement. I don't know whether it's an economic or a coding issue, can anyone figure it out?

I've changed the .py file such that G becomes a vector:
- par.G = 1.03 * np.ones(par.T)
- par.G[t] in all instances

I've changed the notebook for the simulation such that there is a drop in income and no drift from period 16:
- model.par.G[15]=0.5
- model.par.G[16:]=1.0

This leads to increasing savings until period 14. In period 15 everyone increases their consumption (sim.C), and household consumption jumps with on average 0.386 ( compared to an average increase of 0.014 between period 13 and 14).

It seems like all households are positively surprised, which cannot be the case.Here are the average consumption and permanent income figures over the life cycle:

![Skærmbillede 2023-02-17 kl  16 09 20](https://user-images.githubusercontent.com/111353103/219691602-78de75a7-ac0c-4c0f-8c82-a2628bc29775.png)

![Skærmbillede 2023-02-17 kl  16 10 09](https://user-images.githubusercontent.com/111353103/219691781-e14b4f83-20b2-473c-8f80-46874c117445.png)

**Comment by ThomasHJorgensen** [comment]
Does the simulated permanent income look like expected?

Den 17. feb. 2023 kl. 16.11 skrev AntonSteino ***@***.***>:

﻿

Hi, I've looked at the lecture 2 notebook, playing around to introduce a handheld "retirement" period by giving G a time subscript. This results in a strange overconsumption pattern in the last period pre retirement. I don't know whether it's an economic or a coding issue, can anyone figure it out?

I've changed the .py file such that G becomes a vector:

  *   par.G = 1.03 * np.ones(par.T)
  *   par.G[t] in all instances

I've changed the notebook for the simulation such that there is a drop in income and no drift from period 16:

  *   model.par.G[15]=0.5
  *   model.par.G[16:]=1.0

This leads to increasing savings until period 14. In period 15 everyone increases their consumption (sim.C), and household consumption jumps with on average 0.386 ( compared to an average increase of 0.014 between period 13 and 14).

It seems like all households are positively surprised, which cannot be the case.Here are the average consumption and permanent income figures over the life cycle:

[Skærmbillede 2023-02-17 kl  16 09 20]<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F111353103%2F219691602-78de75a7-ac0c-4c0f-8c82-a2628bc29775.png&data=05%7C01%7Cthomas.h.jorgensen%40econ.ku.dk%7C40bfa1718b2842fafc8c08db10f93834%7Ca3927f91cda14696af898c9f1ceffa91%7C0%7C0%7C638122434795560631%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=1DGoUClx1nxbNrJbTMjxyLq0yDF6Spbg2IF0qYVWtSE%3D&reserved=0>

[Skærmbillede 2023-02-17 kl  16 10 09]<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F111353103%2F219691781-e14b4f83-20b2-473c-8f80-46874c117445.png&data=05%7C01%7Cthomas.h.jorgensen%40econ.ku.dk%7C40bfa1718b2842fafc8c08db10f93834%7Ca3927f91cda14696af898c9f1ceffa91%7C0%7C0%7C638122434795560631%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=%2BfgO6Vf9sZg6%2BNjHpv8zwyMpNuskOFmvm%2BFQdeXReV4%3D&reserved=0>

**Comment by AntonSteino** [comment]
Yes, everything else looks as expected - Permanent income, asset accumulation, and transformations.

**Comment by AntonSteino** [comment]
Yes, everything else looks as expected - Permanent income, asset accumulation, and transformations.

**Comment by AntonSteino** [comment]
Yes, both permanent income and asset accumulation look like expected. But there is a strange jump in cash-on-hand. In period 15 average M is much higher than average (1+r)*A+Y, which is why I suspect there is a code error somewhere.

**Comment by AntonSteino** [comment]
Okay, so I've found some bits of the code in the .py file that seem to be wrong. 

In the simulation part, lines 185-6, all time subscripts are 't', although from the equations on the slides, it should have been 't+1' for sim.psi and sim.m. This means that sim.m will be too high in period 15, because the 'fac' will be too small. However, if you try to run the corrected code you get an error, as there is no values for 't = 20'. You could probably shorten the simulation period to par.T-1, but this is more cumbersome than one would think, so I haven't worked that out in full.

Also, line 197 should be 'sim.C = sim.c*sim.P' and not 'sim.C = sim.C*sim.P'.

**Comment by ThomasHJorgensen** [comment]
Thank you very much.

Have look now. I have updated the simulation to be in levels. It should be easy for you to modify the growth rate.

Thomas


Thomas Jørgensen
Associate Professor

University of Copenhagen
Department of Economics

DIR +45 42764004
www.tjeconomics.com<http://www.tjeconomics.com/>

[Logo for Københavns Universitet]

How we protect personal data<https://informationssikkerhed.ku.dk/english/protection-of-information-privacy/privacy-policy/>
